File: mobile/src/main/java/treehou/se/habit/ui/servers/ServersFragment.java
package treehou.se.habit.ui.servers;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.trello.rxlifecycle.components.support.RxFragment;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

import butterknife.OnClick;
import butterknife.Unbinder;
import io.realm.Realm;
import io.realm.RealmResults;
import rx.functions.Action1;
import treehou.se.habit.HabitApplication;
import treehou.se.habit.R;
import treehou.se.habit.core.db.model.ServerDB;
import treehou.se.habit.module.ApplicationComponent;
import treehou.se.habit.ui.inbox.InboxListFragment;
import treehou.se.habit.ui.adapter.ServersAdapter;
import treehou.se.habit.ui.bindings.BindingsFragment;
import treehou.se.habit.util.Settings;

public class ServersFragment extends RxFragment {

    private static final String TAG = "ServersFragment";

    private ViewGroup container;
    @BindView(R.id.list) RecyclerView lstServer;
    @BindView(R.id.empty) View viwEmpty;
    @BindView(R.id.fab_add) FloatingActionButton fabAdd;

    @Inject Settings settings;

    private ServersAdapter serversAdapter;
    private RealmResults<ServerDB> servers;
    private Unbinder unbinder;

    private Realm realm;

    public static ServersFragment newInstance() {
        ServersFragment fragment = new ServersFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    public ServersFragment() {}

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getApplicationComponent().inject(this);
        realm = Realm.getDefaultInstance();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        this.container = container;
        View rootView = inflater.inflate(R.layout.fragment_servers, container, false);
        unbinder = ButterKnife.bind(this, rootView);

        GridLayoutManager gridLayoutManager = new GridLayoutManager(getActivity(), 1);
        lstServer.setLayoutManager(gridLayoutManager);
        lstServer.setItemAnimator(new DefaultItemAnimator());
        setupActionbar();

        return rootView;
    }

    protected ApplicationComponent getApplicationComponent() {
        return ((HabitApplication) getContext().getApplicationContext()).component();
    }

    @Override
    public void onResume() {
        super.onResume();
        setupAdapter();
    }

    @Override
    public void onPause() {
        super.onPause();
        lstServer.setAdapter(null);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        realm.close();
    }

    /**
     * Hookup actionbar
     */
    private void setupActionbar(){
        ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if(actionBar != null) {
            actionBar.setTitle(R.string.servers);
        }
        setHasOptionsMenu(true);
    }

    /**
     * Hookup server list, listening for server updates.
     */
    private void setupAdapter(){
        realm.where(ServerDB.class).findAllAsync().asObservable()
                .compose(this.bindToLifecycle())
                .subscribe(servers1 -> {
                    Log.d(TAG, "Loaded " + servers1.size() + " servers");
                    ServersFragment.this.servers = servers1;
                    updateEmptyView(servers1.size());
                    serversAdapter = new ServersAdapter(servers1);
                    serversAdapter.setItemListener(serverListener);
                    lstServer.setAdapter(serversAdapter);
                });

        if(!settings.getServerSetupAsked()){
            showScanServerFlow();
        }
    }

    /**
     * Launch flow for opening server scanning.
     */
    private void showScanServerFlow() {
        settings.setServerSetupAsked(true);
        new AlertDialog.Builder(getActivity())
                .setMessage(R.string.start_scan_question)
                .setNeutralButton(R.string.new_server, (dialog, which) -> startNewServerFlow())
                .setPositiveButton(R.string.scan, (dialog, which) -> openServerScan())
                .setNegativeButton(R.string.cancel, null)
                .show();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.servers, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_scan_for_server:
                openServerScan();
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    private ServersAdapter.ItemListener serverListener = new ServersAdapter.ItemListener() {
        @Override
        public void onItemClickListener(ServersAdapter.ServerHolder serverHolder) {
            final ServerDB server = serversAdapter.getItem(serverHolder.getAdapterPosition());
            openServerPage(server);
        }

        /**
         * Open page for editing server.
         * @param server the server to open page for.
         */
        private void openServerPage(ServerDB server){
            getActivity().getSupportFragmentManager().beginTransaction()
                    .replace(R.id.page_container, ServerMenuFragment.newInstance(server.getId()))
                    .addToBackStack(null)
                    .commit();
        }

        @Override
        public boolean onItemLongClickListener(final ServersAdapter.ServerHolder serverHolder) {

            final ServerDB server = serversAdapter.getItem(serverHolder.getAdapterPosition());
            showRemoveDialog(serverHolder, server);
            return true;
        }
    };

    /**
     * Launch flow for creating new server.
     */
    @OnClick({R.id.empty, R.id.fab_add})
    public void startNewServerFlow(){
        getActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.page_container, SetupServerFragment.newInstance())
                .addToBackStack(null)
                .commit();
    }

    /**
     * Launch flow used to scan for server on network.
     */
    private void openServerScan(){
        getActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.page_container, ScanServersFragment.newInstance())
                .addToBackStack(null)
                .commit();
    }

    /**
     * Launchs flow asking user to remove or keep server.
     * @param serverHolder holder that triggered flow.
     * @param server the server to remove.
     */
    private void showRemoveDialog(final ServersAdapter.ServerHolder serverHolder, final ServerDB server){
        new AlertDialog.Builder(getActivity())
                .setMessage(R.string.remove_server_question)
                .setPositiveButton(R.string.ok, (dialog, which) -> {
                    realm.beginTransaction();
                    int position = serverHolder.getAdapterPosition();
                    int i = servers.indexOf(server);
                    if(i >= 0) { servers.deleteFromRealm(i);}
                    realm.commitTransaction();
                    serversAdapter.notifyItemRemoved(position);
                })
                .setNegativeButton(R.string.cancel, null)
                .show();
    }

    /**
     * Show empty view if no controllers exist
     */
    private void updateEmptyView(int itemCount){
        viwEmpty.setVisibility(itemCount <= 0 ? View.VISIBLE : View.GONE);
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/servers/serverlist/ServersFragment.kt
package treehou.se.habit.ui.servers.serverlist

import android.app.AlertDialog
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.reactivex.android.schedulers.AndroidSchedulers
import io.realm.RealmResults
import kotlinx.android.synthetic.main.fragment_servers_universal.*
import treehou.se.habit.HabitApplication
import treehou.se.habit.R
import treehou.se.habit.core.db.model.ServerDB
import treehou.se.habit.dagger.ApplicationComponent
import treehou.se.habit.dagger.HasActivitySubcomponentBuilders
import treehou.se.habit.dagger.fragment.ServersComponent
import treehou.se.habit.dagger.fragment.ServersModule
import treehou.se.habit.mvp.BaseDaggerFragment
import treehou.se.habit.ui.adapter.ServersAdapter
import treehou.se.habit.ui.servers.ServerMenuFragment
import treehou.se.habit.ui.servers.create.CreateServerActivity
import treehou.se.habit.util.Settings
import javax.inject.Inject

class ServersFragment : BaseDaggerFragment<ServersContract.Presenter>(), ServersContract.View {

    @Inject lateinit var settings: Settings

    @Inject lateinit var serverPresenter: ServersContract.Presenter

    private lateinit var serversAdapter: ServersAdapter
    private var servers: RealmResults<ServerDB>? = null

    protected val applicationComponent: ApplicationComponent
        get() = (context!!.applicationContext as HabitApplication).component()

    private val serverListener = object : ServersAdapter.ItemListener {
        override fun onItemClickListener(serverHolder: ServersAdapter.ServerHolder) {
            val server = serversAdapter.getItem(serverHolder.adapterPosition)
            if (server != null) {
                openServerPage(server)
            }
        }

        /**
         * Open page for editing server.
         * @param server the server to open page for.
         */
        private fun openServerPage(server: ServerDB) {
            activity!!.supportFragmentManager.beginTransaction()
                    .replace(R.id.page_container, ServerMenuFragment.newInstance(server.id))
                    .addToBackStack(null)
                    .commit()
        }

        override fun onItemLongClickListener(serverHolder: ServersAdapter.ServerHolder): Boolean {

            val server = serversAdapter.getItem(serverHolder.adapterPosition)
            if (server != null) {
                showRemoveDialog(serverHolder, server)
            }
            return true
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        (activity!!.application as HabitApplication).component().inject(this)
        launchScanDialogIfNoServers()
    }

    override fun getPresenter(): ServersContract.Presenter? {
        return serverPresenter
    }

    override fun injectMembers(hasActivitySubcomponentBuilders: HasActivitySubcomponentBuilders) {
        (hasActivitySubcomponentBuilders.getFragmentComponentBuilder(ServersFragment::class.java) as ServersComponent.Builder)
                .fragmentModule(ServersModule(this))
                .build().injectMembers(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        return inflater.inflate(R.layout.fragment_servers_universal, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
        serverList.layoutManager = layoutManager
        serverList.itemAnimator = DefaultItemAnimator()
        serversAdapter = ServersAdapter()
        serversAdapter.setItemListener(serverListener)
        serverList.adapter = serversAdapter

        setupActionbar()

        addFab.setOnClickListener { startNewServerFlow() }
        emptyText.setOnClickListener { startNewServerFlow() }
    }

    override fun onResume() {
        super.onResume()
        setupAdapter()
    }

    override fun onDestroy() {
        super.onDestroy()
        realm.close()
    }

    /**
     * Hookup actionbar
     */
    private fun setupActionbar() {
        val actionBar = (activity as AppCompatActivity).supportActionBar
        actionBar?.setTitle(R.string.servers)
    }

    private fun launchScanDialogIfNoServers() {
        realm.where(ServerDB::class.java).findAllAsync().asFlowable().toObservable()
                .compose(this.bindToLifecycle<RealmResults<ServerDB>>())
                .observeOn(AndroidSchedulers.mainThread())
                .filter({ it.isLoaded })
                .firstElement()
                .subscribe({
                    if (it.size == 0) {
                        showScanServerFlow()
                    }
                }, { logger.e(TAG, "launchScanDialogIfNoServers", it) })
    }

    /**
     * Hookup server list, listening for server updates.
     */
    private fun setupAdapter() {
        realm.where(ServerDB::class.java).findAllAsync().asFlowable()
                .filter({ it.isLoaded && it.isValid })
                .compose(this.bindToLifecycle<RealmResults<ServerDB>>())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    Log.d(TAG, "Loaded " + it.size + " servers")

                    this@ServersFragment.servers = it
                    updateEmptyView(it.size)
                    serversAdapter.setItems(it.createSnapshot())
                }, { logger.e(TAG, "Failed to load servers", it) })
    }

    /**
     * Launch flow for opening server scanning.
     */
    private fun showScanServerFlow() {
        AlertDialog.Builder(activity)
                .setMessage(R.string.start_scan_question)
                .setPositiveButton(R.string.new_server) { _, _ -> startNewServerFlow() }
                .setNegativeButton(R.string.cancel, null)
                .show()
    }

    /**
     * Launch flow for creating new server.
     */
    fun startNewServerFlow() {
        startNewServerReveal()
    }

    private fun startNewServerReveal() {
        startNewServerRevealBasic()
    }

    private fun startNewServerRevealBasic() {
        val activity = activity
        if (activity != null) {
            val intent = CreateServerActivity.createIntent(activity)
            startActivity(intent)
        }
    }

    /**
     * Launchs flow asking user to remove or keep server.
     * @param serverHolder holder that triggered flow.
     * @param server the server to remove.
     */
    private fun showRemoveDialog(serverHolder: ServersAdapter.ServerHolder, server: ServerDB) {
        AlertDialog.Builder(activity)
                .setMessage(R.string.remove_server_question)
                .setPositiveButton(R.string.ok) { _, _ ->
                    realm.beginTransaction()
                    val position = serverHolder.adapterPosition
                    val i = servers!!.indexOf(server)
                    if (i >= 0) {
                        servers!!.deleteFromRealm(i)
                    }
                    realm.commitTransaction()
                    serversAdapter.notifyItemRemoved(position)
                }
                .setNegativeButton(R.string.cancel, null)
                .show()
    }

    /**
     * Show empty view if no controllers exist
     */
    private fun updateEmptyView(itemCount: Int) {
        emptyText.visibility = if (itemCount <= 0) View.VISIBLE else View.GONE
    }

    companion object {

        private val TAG = "ServersFragment"

        fun newInstance(): ServersFragment {
            val fragment = ServersFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}
